README.md
shopebevel
==========


magiczoomplus/CHANGELOG
4.7 (2012-12-20)
    - Module no longer overrides Mage_Catalog_Block_Product_View_Options_Type_Select class.
    - Added thumbnail to switch between Magic 360 and other (Magic Zoom Plus, Magic Zoom, Magic Thumb, Magic Magnify, Magic Magnify Plus, Magic Touch) module view if they are both installed.
    - Separate tab on the settings page of the product for loading Magic 360 (Flash) images.
4.6 (2012-09-07)
    - Added ability to display custom block with images for MagicSlideshow and MagicScroll modules.
4.5 (2012-08-22)
    - Added support for 'Use Categories Path for Product URLs' option.
    - Added ability to show associated product's images when option is selected or within a gallary.
    - Added option: Include headers on all pages (default: No).
    - Improvements in installation process.
    - Improvement in the settings page.
    - Added Recently Viewed block support.
4.4 (2011-06-14)
    - Added templates (replaced selectors-position parameter)
    - Added Magic Scroll for selectors
4.3 (2011-06-14)
    - Add compatibility with Magento compiler.
    - Fix issue with 'isModuleEnabled' function in Magento 1.4.0.1
    - Fix issue with 'thumbnails' option in MagicSlideshow module.
    - Fix some warnings in installer.
4.2 (2011-06-06)
    - Added capability combined usage of magic360(Plus) with other magic modules.
4.1 (2011-06-01)
    - Added installer.
4.0 (2011-04-28)
    - Now supports Magento native modules system.
    - Moved module configuration to Magento admin panel.
    - Code rewrite.
3.36 (2011-02-02)
    - Fixed fatal error in module on reviews page.
3.35 (2011-01-17)
    - Added Magic Scroll support
3.34 (2010-12-22)
    - New option: 'selector-position' (default: bottom).
3.33 (2010-12-15)
    - Added support for Magic360 and Magic360plus
3.32 (2010-07-28)
    - Added support custom option of type "Radio Buttons".
3.31 (2010-07-28)
    - Fixed menu conflict in Magento v 1.4.1.x and above.
3.30 (2010-07-23)
    - Fixed PHP Warning when product image does not exist.
3.29 (2010-07-09)
    - Fixed fatal error on same ajax pages.
3.28 (2010-06-25)
    - Custom option values and image labels are now case-insensitive.
3.27 (2010-06-08)
    - Changes in caption-source parameter to allow multiple sources.
3.26 (2010-05-31)
    - Fixed some locale problems.
3.25 (2010-05-26)
    - Added support for configurable product attributes.
    - Fixed JavaScript error in magictoolbox_utils.js.
3.24 (2010-04-14)
    - Fixed JavaScript error when product has no selectors.
3.23 (2010-03-30)
    - Watermark image available on thumbnails.
3.22 (2010-03-24)
    - Added support for multiple options.
3.21 (2010-03-18)
    - New option switch in Magnify.
3.20 (2010-03-02)
    - Headers no longer load on search results page.
3.19 (2010-02-24)
    - Installer is now compatible with Magento v 1.4.x
3.18 (2010-02-03)
    - Headers only load when needed, not on every page.
    - More options available for Magic Thumb modules.
    - Some code optimizations.
3.17 (2010-02-01)
    - Fixed wrong thumb size on category pages when selector is clicked.
    - Added 'category-selector-size' parameter.
    - Fixed module not working on category page in rare cases.
3.16 (2010-01-13)
    - Magic Zoom module now works on category pages when use_effect_on_category_page is used.
3.15 (2009-12-21)
    - Installer optimizations.
3.14 (2009-12-10)
    - Installer will now check path to themes.
    - Separate config file was working only with default theme.
    - Effect now works on all categories.
    - Extra white space in option price removed from option_associated_with_images option.
3.13 (2009-12-03)
    - Captions problem resolved on category page.
3.12 (2009-11-11)
    - Options now work if captions are off.
3.11 (2009-10-28)
    - Added 'Upgrade from ZIP' option to installer.
3.10 (2009-10-23)
    - Module optimizations to prevent fatal memory limit error in PHP.
3.9 (2009-10-19)
    - Options changed when 'thumb_change' parameter is set to 'mouseover'.
3.8 (2009-09-23)
    - Added support to new.phtml file (the new products page).
3.7 (2009-09-10)
    - Fixed the ignore_magento_css option error.
3.6 (2009-09-01)
    - MagicZoom and MagicThumb class values for now work for use_effect_on_category_page.correctly
    - Added category_thumb_size option.
3.5 (2009-08-31)
    - Fixed Onclick on category pages, while MagicThumb is active.
3.4 (2009-08-28)
    - PHP file permisions error fixed in module installer.
3.3 (2009-08-17)
    - Caption_source option error fixed on category page.
3.2 (2009-08-14)
    - Magento shop available if image file is not avaliable.
3.1 (2009-07-27)
    - Module now works with more Magento themes.
    - Some code optimisation.
3.0 (2009-07-06)
    - Added installer.
    - Solved the W3C validation error. All headers now placed in <head> section.
2.17 (2009-07-01)
    - Supports all templates for category page.
    - Added support for different locales.
2.16 (2009-06-15)
    - Added new parameter: Size of thumbnail depends on.
2.15 (2009-06-12)
    - Product name shows as title when additional image has no individual title.
2.14 (2009-06-11)
    - Remove PHP code from category page (list.phtml file - short PHP tags not allowed by Magento).
2.13 (2009-05-08)
    - New option 'use_individual_titles' (default: Yes).
2.12 (2009-04-21)
    - Added new file 'list_modern.phtml' for modern skin.
2.11 (2009-04-15)
    - New option: 'show_selectors_on_category_page' (default: No).
2.10 (2009-04-03)

js/pixafy/pixvote.js

	//init publish and subscription methodology
	initPubSub();

	$j(function()
	{
		$j("body").on("click", ".view-comments-display", function( e )
		{
			var self = $j(this);

			var container = self.parents( "#comment-area-" + self.data("design-id") ).find( '.comments-list' );

			//show the loader while we process the request
			showLoader( container );

			$j.post(
				getUrl(VOTE_FRONTNAME + "/comment/getcomments"),
				{id:self.data("design-id")},
				buildCommentList,
				"json"
			);

			e.preventDefault();

		});

		$j("body").on("click", ".vote-btn.logged-in.new-vote", function()
		{
			// show vote success message and disable vote button
			// update vote count
			var button = $j( this ),
				alert = button.siblings( '.following-alert' ),
				countContainer = button.siblings( '.vote-count' ),
				count = countContainer.children( 'span' ),
				newCount = parseInt( count.text(), 10 );
				newCount++;

				// fade out the count container and the button
				countContainer.fadeOut( 300 );
				button.fadeOut( 300, function() {
					// remove the vote-btn class to disable and set the inactive class
					button.removeClass( 'vote-btn first-vote' ).addClass( 'grey' );
					// fade in the success message
					alert.fadeIn( 300, function() {
						// pause for 1.5 seconds, then fade out alert
						setTimeout(function() {
							alert.fadeOut( 300, function() {
								// upon completion fade in button, update vote count
								// and fade in vote count container
								button.fadeIn( 300 );
								count.text( newCount );
								countContainer.fadeIn( 300 );

								// a bleet, a bleet, a bleet, that's all folks!!
							});
						}, 1500);
					});
				});


			if(!VOTE_UPDATE_OPERATING)
			{
				//update vote counter
				VOTE_UPDATE_OPERATING = true;
				var data = {total_vote_count: $j('.counter').data('stop')};
				data.total_vote_count++;
				updateCounter(data);
			}

			$j.post(
				getUrl(VOTE_FRONTNAME + "/index/addVote"),
				{id:$j(this).data("design-id")},
				voteResult,
				'json'
			);
		});

		$j("body").on("click", ".comments-login", showLogin);


		$j("body").on("click", ".vote-comment-btn", function()
		{
			//get all the elements we need and the comment and package them
			var btn = $j( this );
			var designId = btn.attr( 'data-design-id' );
			var btns = $j( '.btn-' + designId );
			var inputs = $j( '.input-' + designId );
			var commentLists = $j( '.list-' + designId );

			var obj = {
				designId : designId,
				btns : btns,
				inputs : inputs,
				commentLists : commentLists
			};

			$j.publish( 'comment/bundle/packaged', obj );
		});

	}); //end doc ready


	/* set subscriptions */
	$j.subscribe( 'comment/bundle/packaged', function( e, obj ) {
		prepCommentForSubmit( obj );
	});

	$j.subscribe( 'comment/bundle/prepped', function( e, obj ) {
		submitComment( obj );
	});

	$j.subscribe( 'comment/returned', function( e, obj ) {
		prepCommentList( obj );
	});

	$j.subscribe( 'comment/prepped', function( e, obj ) {
		showCommentList( obj );
	});

	$j.subscribe ( 'update/counts', function( e, obj ) {
		updateCommentCount( obj );
	});

	function prepCommentForSubmit( obj ) {

		//boolean for error handling
		var hasComment = false;

		$j.each( obj.inputs, function() {
			//cache this
			var self = $j( this );

			//get the value of the input
			var val = self.val();

			//init our comment text var
			var commentText = '';

			if ( ( val !== '' ) && ( val !== 'Add a comment...' ) ) {
				commentText = val;

				//clear the value in the comment field
				self.val( '' );

				obj.commentText = commentText;

				showLoaders( obj );

				hasComment = true;
			}
		});

		if ( !hasComment ) {
			window.alert( 'Please add a comment before submitting.' );
		}
	}

	function submitComment( obj ) {

		$j.each( obj.btns, function() {
			$j( this ).removeClass( 'vote-comment-btn' );
		});

		$j.post(
			getUrl(VOTE_FRONTNAME + "/comment/addcomment"),
			{
				'comment':
				{
					'design_id' : obj.designId,
					'comment' : obj.commentText
				}
			},
			addCommentResult,
			'json'
		);
	}

	//success passes back id, comment, image, comment count
	function addCommentResult( data )
	{
		var container = {};

		if(data.hasOwnProperty("error"))
		{
			//cache the container for our comments
			container = $j( "#comment-area-" + data.id ).find( '.comments-list' );

			//if error, show the end user the error of their ways
			var listItem = $j('<li/>', {
				'class'	: 'comments-list-item clear-box',
				text	: data.error
			}).appendTo( container );
			showComments( container );
		}
		else
		{
			//get all the elements we need and package them
			var designId = data.id;
			var btns = $j( '.btn-' + designId );
			var inputs = $j( '.input-' + designId );
			var commentLists = $j( '.list-' + designId );

			var obj = {
				designId : designId,
				btns : btns,
				inputs : inputs,
				commentLists : commentLists,
				data : data //add data to the object
			};

			$j.publish( 'comment/returned', obj );
		}
	}

	function prepCommentList( obj ) {
		$j.each( obj.commentLists, function() {
			buildCommentListItem( obj.data, this );
		});

		$j.publish( 'comment/prepped' , obj );
		$j.publish( 'update/counts', obj );
	}

	function buildCommentListItem( item, container ) {

		//create our list item and append it to our ul
		var listItem = $j('<li/>', {
			'class'	: 'comments-list-item clear-box'
		}).appendTo( container );

		//create our image container div and append it to the list item
		var imageContainer = $j('<div/>', {
			'class'	: 'comment-image'
		}).appendTo(listItem);

		//create our image and append it to the image container
		var image = $j('<img/>', {
			src : item.image
		}).appendTo( imageContainer );

		//create our comment container and append it to the list item
		var commentContainer = $j('<div/>', {
			'class'	: 'comment-text'
		}).appendTo( listItem );

		//create our author span and append it to the comment container
		var author = $j('<span/>', {
			'class'	: 'author',
			text	: item.author + ' '
		}).appendTo( commentContainer );

		//create our comment span and append it to the comment container
		var comment = $j('<span/>', {
			'class'	: 'text',
			text	: item.comment
		}).appendTo( commentContainer );
	}

	function buildCommentList( data ) {

		var container = {};

		//cache the container for our comments
		container = $j( '.comments-list.see-all' );

		container.empty();

		$j.each(data.comments, function() {
			buildCommentListItem( this, container);
		});

		showComments( container );

		container.jScrollPane();
	}

	function showCommentList( obj ) {
		//remove loaders from comment list parents
		$j.each( obj.commentLists, function() {
			removeLoader( this );
		});

		//show the containers
		$j.each( obj.commentLists, function() {
			showComments( this );
		});

		//add class back to submit button
		$j.each( obj.btns, function() {
			$j( this ).addClass( 'vote-comment-btn' );
		});
	}

	function showComments( el ) {
		$j( el ).fadeIn( 300 );
	}

	function updateCommentCount( obj ) {
		var countBoxes = $j( '.count-box-' + obj.designId );
		var count = obj.data.count;

		$j.each( countBoxes, function() {
			$j( this ).text( count );
		});
	}

	function showLoaders( obj ) {

		$j.each( obj.commentLists, function() {
			var self = $j( this );
			var parent = self.parent();

			self.hide();
			self.empty();

			$j('<div>', {
				'class' : 'comments-loader'
			}).prependTo( parent );
		});

		$j.publish( 'comment/bundle/prepped', obj );
	}

	function showLoader( container ) {

		var self = $j( this );
		var parent = self.parent();

		self.hide();
		self.empty();

		$j('<div>', {
			'class' : 'comments-loader'
		}).prependTo( parent );

	}

	function removeLoader( container ) {
		$j( container ).siblings( '.comments-loader' ).remove();
	}

	//sucess returns id,vote_count
	function voteResult(data)
	{
		var container = {};

		if(data.hasOwnProperty("error"))
		{
			// var message = '<?php echo $this->getLayout()->createBlock("cms/block")->setBlockId("submit-vote-login")->toHtml() ?>';
			// $j( '.comment-input' ).on( 'focus', function() {
			// 	showLogin( message );
			// });
		}
		else
		{
            $j(document).trigger('shopBevel_designVoted', data.id);
			if(data.hasOwnProperty("store_credit"))
			{
				$j("#credit-amt").html(data.store_credit);
			}
			//get our vote count container and update the value by one
			var voteCountContainer =	$j('p#vote-' + data.id);
			voteCountContainer.text( data.vote_count );
			//checkForVotes();
			$j('.counter').queue(function(next){
				if(!VOTE_UPDATE_OPERATING)
				{
					VOTE_UPDATE_OPERATING = true;
					updateCounter(data);
				}
				next();
			});
		}
	}


	function shareVoteResult(data)
	{
		var container = {};

		if(data.hasOwnProperty("error"))
		{
			// var message = '<?php echo $this->getLayout()->createBlock("cms/block")->setBlockId("submit-vote-login")->toHtml() ?>';
			// $j( '.comment-input' ).on( 'focus', function() {
			// 	showLogin( message );
			// });
		}
		else
		{
			ajaxVoteUpdate(data);
			//$j.post(getUrl(VOTE_FRONTNAME + "/index/addCredit"));
			// dismiss the current modal
			$j( '.pixafy-modal-dismiss' ).trigger( 'click' );

			// show thank you for sharing modal
			trackPopup($j("#popup_thanks_subject").html(), $j("#popup_thanks_version").html());
			trackSuccess($j("#popup_share_subject").html());
			setTimeout(function() {
				$j( '#share-double-thanks-trigger' ).trigger( 'click' );
			}, 1000);

		}
	}

	function ajaxVoteUpdate(data)
	{
		$j('.counter').queue(function(next){
			if(!VOTE_UPDATE_OPERATING)
			{
				VOTE_UPDATE_OPERATING = true;
				updateCounter(data);
			}
			next();
		});
	}

	function trackPopup(trackSubject, trackVersion)
	{
		$j.post(
					getUrl(VOTE_FRONTNAME + "/index/track"),
					{
						subject:trackSubject,
						version:trackVersion
					}
				);
	}

	function trackSuccess(trackSubject)
	{
		$j.post(
					getUrl(VOTE_FRONTNAME + "/index/trackSuccess"),
					{
						subject:trackSubject
					}
				);
	}

	function setVoteItems( items ) {

		$j.each( items, function() {
			var self =$j( this );
			if ( self.hasClass( 'no-right-margin' ) ) {
				self.removeClass('no-right-margin');
			}
		});

		$j( '#vote-item-collection' ).columnsUpdate( {}, items );
	}


	//Function to update pagination vars after ajax call
	function updatePaginationVars()
	{
		OPERATING = false;
	}

	//Function for top section next/prev button
	$j(function()
	{
		$j("body").on("click", "#next-btn", function(){
			if(FEATURED_PAGE < FEATURED_TOTAL)
			{
				var container = $j( '.vote-display-container' );
				container.css( 'height', 432 + 'px' ).children().fadeOut( 300 );
				$j('<div>', {
					'class' : 'featured-loader'
				}).appendTo(container);
				featuredPaginate("next");
			}
		});
		$j("body").on("click", "#prev-btn", function(){
			if(FEATURED_PAGE > 1)
			{
				var container = $j( '.vote-display-container' );
				container.css( 'height', 432 + 'px' ).children().fadeOut( 300 );
				$j('<div>', {
					'class' : 'featured-loader'
				}).appendTo(container);
				featuredPaginate("prev");
			}
		});
	});

	function featuredPaginate(direction)
	{
		switch(direction)
		{
			case "next":
			FEATURED_PAGE++;
			break;
			case "prev":
			FEATURED_PAGE--;
			break;
			default:
			break;
		}
		$j.post(FEATURED_PAGEINATION_URL, {page:FEATURED_PAGE}, featuredPaginateResult, 'json');
	}

	function featuredPaginateResult(data)
	{
		$j(".vote-display-container").html(data.html).css('height', 'auto');
	}

	//Function for pagination init
	function setPaginationVars(perPage, startingPage, total, url)
	{
		PER_PAGE = perPage;
		PAGE = startingPage;
		TOTAL = total;
		PAGINATION_URL = url;
	}

	function checkScrollBottom()
	{
		//var target = window;
		var target = ".footer-container";
		var curHeight =  $j(window).scrollTop() + $j(window).height();
		return curHeight >= $j(target).height() || $j(target).height() - curHeight == 1;
	}


	//Function to get next set of designs
	function ajaxPaginate()
	{
		OPERATING = true;
		PAGE++;
		$j.post(PAGINATION_URL, {page:PAGE}, ajaxPaginateResult, 'json');
	}

	//Process result from pagination call
	function ajaxPaginateResult(data)
	{
		var items = $j(data.html).find(".product-designer");
		$j.each( items, function() {

			var self =	$j( this ),
				id =	self.find( '.vote-info' )
							.children( 'button' )
							.attr( 'data-design-id' ),
				image =	self.find( '.prdt-img' )
							.children( 'img' )
							.first(),
				link =	self.find( '.vote-item-share' )
							.attr( 'data-url' ),
				name =	self.find( '.vote-dsgnr-desc' )
							.find( '.txt-dsgnr-name' )
							.text(),
				src =	'';

				if ( image.length > 0 ) {
					src = image.attr( 'src' );
				}

				VOTE_ITEM_INFO[ id ] = {
					name: name,
					url: link,
					image: src
				};
		});
		setVoteItems( items );
		updatePaginationVars();
	}

	//Function to update pagination vars after ajax call
	function updatePaginationVars()
	{
		OPERATING = false;
	}


//Function for pagination init
	function setPaginationVars(perPage, startingPage, total, url)
	{
		PER_PAGE = perPage;
		PAGE = startingPage;
		TOTAL = total;
		PAGINATION_URL = url;
	}

	function checkScrollBottom()
	{
		//var target = window;
		var target = ".footer-container";
		var curHeight =  $j(window).scrollTop() + $j(window).height();
		return curHeight >= $j(target).height() || $j(target).height() - curHeight == 1;
	}

	//Function to get next set of designs
	function ajaxPaginate()
	{
		OPERATING = true;
		PAGE++;
		$j.post(PAGINATION_URL, {page:PAGE}, ajaxPaginateResult, 'json');
	}

	function initPubSub() {
		/**
		 * Create an observer pattern for jQuery
		 * allows you to alias trigger with publish,
		 * on with subscribe,
		 * and off with unsubcsribe
		 * Useful for custom events
		 *
		 *USAGE:
		 * $(el).subscribe('customEvent', function() {
		 *       do something when customEvent is published
		 *
		 *       No longer need to subscribe to custom event
		 *       $(el).unsubscribe('customEvent');
		 * });
		 *
		 * Custom event is fired
		 * $.publish('customEvent');
		 */

		var o = $j( {} );
		$j.each({
			trigger: 'publish',
			on: 'subscribe',
			off: 'unsubscribe'
		}, function( key, val ){
			jQuery[ val ] = function() {
				o[ key ].apply( o, arguments );
			};
		});
	}

	/* set up functionality for sharing a vote on facebook */
	(function( $ ) {
		$(function() {

			/* instanitate share modals */
			$( '#share-double-modal' ).pixafyModal({
				transition: 'fade',
				keepOverlay: 'true',
				enter: function() {
					$( '.pixafy-modal-overlay' ).addClass( 'share-overlay' );
				},
				exit: function() {
					// shows store credit with an attention drawing animation
					setTimeout(function() {
						bounceCredit();
					}, 500);
					$( '.pixafy-modal-overlay' ).removeClass( 'share-overlay' );
				}
			});

			$( '#share-double-thanks-modal' ).pixafyModal({
				transition: 'fade',
				noOverlay: 'true'
			});

			$( '#share-double-no-thanks-modal' ).pixafyModal({
				transition: 'fade',
				noOverlay: 'true'
			});

			/* set up our event handlers */
			$( 'body' ).on( 'click', '.vote-btn.first-vote.logged-in', function() {
				/*
				 * when an item is voted pass the item id to the share modal so
				 * we can share the right information
				 */
				var self = $( this );
				var shareModal = $( '#share-double-modal' );
				var id = self.attr( 'data-design-id' );

					shareModal.attr( 'data-elem', id );

					shareModal.find( '.profile-pic' )
						.attr( 'src', VOTE_ITEM_INFO[ id ].designer_image );

					shareModal.find( '.profile-name' )
						.text( VOTE_ITEM_INFO[ id ].designer_name );

					shareModal.find( '.profile-loc' )
						.text( VOTE_ITEM_INFO[ id ].designer_loc );


					//ajax post to create tracking record
					//j("#popup_share_subject").html()
					trackPopup($("#popup_share_subject").html(), $("#popup_share_version").html());
					
					$( '#share-double-trigger' ).trigger( 'click' );
			});

			$( 'body' ).on( 'click', '.pixafy-modal-overlay.share-overlay', function() {
				trackPopup($("#popup_no_thanks_subject").html(), $("#popup_no_thanks_version").html());
				setTimeout(function() {
					$( '#share-double-no-thanks-trigger' ).trigger( 'click' );
				}, 1000);
			});

			$( '#no-share-btn' ).on( 'click', function() {
				$( '.pixafy-modal-dismiss' ).trigger( 'click' );
				trackPopup($("#popup_no_thanks_subject").html(), $("#popup_no_thanks_version").html());
				setTimeout(function() {
					$( '#share-double-no-thanks-trigger' ).trigger( 'click' );
				}, 1000);
			});

			$( 'body' ).on( 'click', '.fb-share-double-btn', function( e ) {
				/*
				 * when the fb share button is clicked prep our FB.ui call with
				 * the proper share info
				 */
				shareVote();
				e.preventDefault();
			});
		}); //end doc ready

		function shareVote() {
			var	modal = $( '#share-double-modal' ),
				id = modal.attr( 'data-elem' ),
				image = VOTE_ITEM_INFO[ id ].image,
				link = VOTE_ITEM_INFO[ id ].url,
				name = VOTE_ITEM_INFO[ id ].item_name;

			FB.ui({
				method: 'feed',
				name: name,
				link: link,
				picture: image,
				caption: 'I voted to produce this: Vote too!',
				description: $j( 'meta[property="og:description"]' ).attr( 'content' )
			},
			function( response ) {
				if ( response && response.post_id ) {
					$j.post(
						getUrl(VOTE_FRONTNAME + "/index/addVote"),
						{"id":id, "shareType":"fb"},
						shareVoteResult,
						'json'
					);

				} else {
					// ?? need direction on error handling ??
				}
			});
		}

		function bounceCredit() {
			// get the store credit container and show it
			var credit = $( '.account-container' ).find( '.store-credit' );

			if ( credit.css( 'display' ) !== 'block' || 'inline' ) {
				credit.fadeIn( 300 );

				// once it is visible, bounce it!
				(function bounce( index, direction ) {
					index++;

					if( index !== 5 ) {
						switch ( direction ) {
							case 'up':
								credit.animate({
									'bottom' : 27 + 'px'
								},300, function() {
									bounce( index, 'down' );
								});
							break;

							case 'down':
								credit.animate({
									'bottom' : 18 + 'px'
								},300, function() {
									bounce( index, 'up' );
								});
							break;
						}
					}
				})( 0, 'up' );
			}
		}
	})( jQuery );

